README.md
Some classes to read and write properties files. 

Properties files are like Java Properties file. The implemented format is using
to store locales for an application made with [Jelix](https://jelix.org), 
a PHP Framework.

installation
============

You can install it from Composer. In your project:

```
composer require "jelix/properties-file"
```

Usage
=====

You have two classes: `Properties` which is a container for key/value pairs,
a `Reader` to  parse a properties file and a `Writer` to write properties into
a file.

```php

use \Jelix\PropertiesFile\Properties;
use \Jelix\PropertiesFile\Parser;
use \Jelix\PropertiesFile\Writer;

$properties = new Properties();

$reader = new Parser();
$reader->parseFromFile('file.properties', $properties);

$value = $properties->get('a_key');
$value = $properties['a_key'];

$properties->set('a_key', 'new_value');
$properties['a_key'] = 'new_value';

$writer = new Writer();
$writer->writeToFile($properties, 'file.properties');

// with a limit of line length (default is 120)
$writer->writeToFile($properties, 'file.properties', 
                    array("lineLength"=>80));

```

Options for the writer: 

- `lineLength`: maximum length of a line. If the string length is higher, it will be splitted. (default: 120)
- `cutOnlyAtSpace`: to cut lines where there is space, not on the middle of a word (default: true) 
- `spaceAroundEqual`: to add or not space around the equal sign (boolean, default: true)
- `headerComment`: to add comment as header (string, default: empty string)
- `removeTrailingSpace`: to remove trailing space on values (boolean, default: false)
- `encoding`: encoding of values to write. (string, default: "UTF-8")

History
=======

The parser is based on a class, jBundle coming from the [Jelix Framework](http://jelix.org)
until Jelix 1.6, and has been released in 2018 into a separate repository as Jelix\PropertiesFile\Parser.

Format
======

The file content structure is quite simple. It's basically a `key=string`
structure, with some improvements. 

You can't use double and single quotes to delimit your strings, new lines do this.

Keys can contain characters `a` to `z` (lowercase/uppercase), numbers and
characters `_`, `-`, `.`.

Here is an example of file:

<code ini>
title.offlineElements = elements to check
title.onlineElements = online elements
buttons.save = Save
buttons.ok=Ok
</code>

Multi line
----------

If the text is long and you want to write it in several lines, you can type an
anti-slash `\` at the end of each line (excepted the last one of the text), to tell
the parser to continue reading the translated string.

<code ini>
intro=this is a very very\
long text in\
several lines
message=this is a regular line
</code>

However, it doesn't insert a line break in the displayed string. If you want to
insert a real line break, use `\n` or `\r` (`\r\n` on windows, `\n` on linux, 
`\r` on macs):

<code ini>
intro=this is a very very \
long text in\nseveral lines, but in\n one line\nin the source
</code>

Comments
--------

You can also put some comments. They have to begin with a `#`. When the parser sees
`#`, the rest of the line is ignored. A comment can be at the beginning of a line,
or in the middle of a line, or at the end of the line. If you want to use a `#` in a
value, you have to escape it with an anti-slash: `\#`.

Whitespaces
-----------

Whitespaces before and after a value are ignored. If you want to put a value equal
to a space, you have to use `\s`.

<code ini>
nospace= #this is using a regular space
space= \s#this is using a \s space
</code>

The value of `space` will be `' '`, and the value of `nospace`, an empty string.

You can also use `\S` to insert an 'unbreakable' space.


CHANGELOG.md
Changelog
==========

1.2.2 (2022-03-15)
------------------

Fix an issue with PHP 8.1

1.2.1 (2019-07-31)
------------------

- fix split of long line and support of encoding

1.2 (2019-07-29)
----------------

- new option for the writer: `cutOnlyAtSpace`.
  Set it to false to cut line by ignoring words


1.1 (2018-10-12)
----------------

- new options for the writer:
    - `spaceAroundEqual`: to add or not space around the equal sign (boolean, default: true)
    - `headerComment`: to add comment as header (string, default: empty string)
    - `removeTrailingSpace`: to remove trailing space on values (boolean, default: false)

1.0 (2018-10-11)
----------------

Initial release



tests/parserTest.php
<?php
/**
 * @author      Laurent Jouanneau
 * @copyright   2018 Laurent Jouanneau
 * @link        http://jelix.org
 * @licence     MIT
 */
use Jelix\PropertiesFile\Parser;
use Jelix\PropertiesFile\Properties;

class parserTest extends \PHPUnit\Framework\TestCase
{
    public function testUnknownFile(){
        try {
            $parser = new Parser ();
            $prop = new Properties();
            $parser->parseFromFile(__DIR__.'/unknown.properties', $prop);
            self::fail('should throw an exception when trying reading unknownfile');
        }catch(Exception $e){
            $this->assertEquals('Cannot load the properties file '.__DIR__.'/unknown.properties', $e->getMessage(),
                'should throw the right exception when trying reading unknownfile (wrong message: '.$e->getMessage().')');
        }
    }

    public function getPropertiesContent(){
        return array(
            array('test_A.properties', []),
            array('test_B.properties', array("aaa"=>"bbb","ccc"=>"")),
            array('test_C.properties', array("aaa"=>"bbb","ccc"=>"ddd")),
            array('test_D.properties', array("module.description"=>"Tests\nunitaires jelix")),
            array('test_E.properties', array("module.description"=>"Tests\ \\\\unitaires jelix")),
            array('test_F.properties', array("module.description"=>"Tests unitaires jelix" )),
            array('test_G.properties', array(
                "module.description" => "Tests unitaires jelix",
                "ooo" => "bbbb",
                "bbb" => " ",
                "ddd" => "lorem ipsum &#65; <html> &quote; test &gt;",
                "ee" => " ",
                "ff" => "  # other",
                "hh" => "    ",
                "ii" => "   ".utf8_encode(chr(160)).' bidule',
                "jj" => "truc"
            )),
            array('test_H.properties', array("module.description" => "Tests unitaires # jelix", "ooo" => "bbbb",)),
            array('test_I.properties',  array("module.description" => "Tests unitaires # jelix",
                "ooo" => " bb bb  ",)),
            array('test_J.properties', array(
                "text.key" => "bug 639 there shouldn't have a notice during the parsing of this property ",
                "text.key2" => "same problem but with spaces at the end of the last line ",
                "text.key3" => "youpa",
            ))
        );
    }

    /**
     * @dataProvider getPropertiesContent
     */
    public function testParserFiles($file, $expected){
        $parser = new Parser ();
        $props = new Properties();
        $parser->parseFromFile(__DIR__.'/assets/'.$file, $props);

        $strings = $props->getAllProperties();
        $this->assertEquals($expected, $strings);
    }

}

tests/writerTest.php
<?php
/**
 * @author      Laurent Jouanneau
 * @copyright   2018 Laurent Jouanneau
 * @link        http://jelix.org
 * @licence     MIT
 */
use Jelix\PropertiesFile\Writer;
use Jelix\PropertiesFile\Properties;

class writerTest extends \PHPUnit\Framework\TestCase
{
    public function getPropertiesContent(){
        return array(
            array(
                array(
                    "aaa"=>"bbb",
                    "ccc"=>""
                ),
                "aaa = bbb\n".
                "ccc = \n",
                array("lineLength"=>80)
            ),
            array(
                array(
                    "aaa"=>"Lorem #Ipsum is\nsimply dummy text ",
                    "ccc"=>"ddd"
                ),
                "aaa = Lorem \\#Ipsum is\\nsimply dummy text\\s\n".
                "ccc = ddd\n",
                array("lineLength"=>80)
            ),
            array(
                array(
                    "html" => "lorem <ipsum>&#65; <html> &quote; test &gt;",
                    "ee" => " ",
                    "ff" => "  # other",
                    "hh" => "    ",
                    "ii" => "   ".utf8_encode(chr(160)).' bidule',
                    "jj" => "truc "
                ),
                "html = lorem <ipsum>&\\#65; <html> &quote; test &gt;\n".
                "ee = \\s\n".
                "ff = \\s \\# other\n".
                "hh = \\s  \\s\n".
                "ii = \\s  \\S bidule\n".
                "jj = truc\\s\n",
                array("lineLength"=>80)
            ),
            array(
                array(
                    "html" => "lorem <ipsum>&#65; <html> &quote; test &gt;",
                    "ee" => " ",
                    "ff" => "  # other",
                    "hh" => "    ",
                    "ii" => "   ".utf8_encode(chr(160)).' bidule',
                    "jj" => "truc "
                ),
                "html = lorem <ipsum>&\\#65; <html> &quote; test &gt;\n".
                "ee = \n".
                "ff = \\s \\# other\n".
                "hh = \n".
                "ii = \\s  \\S bidule\n".
                "jj = truc\n",
                array("lineLength"=>80, "removeTrailingSpace" => true)
            ),
            array(
                array(
                    "long.description" => "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    "exact.max.length" => "Lorem ipsum"),
                "long.description = Lorem ipsum\\\n".
                "dolor sit amet, consectetur ad\\\n".
                "ipiscing elit.\n".
                "exact.max.length = Lorem ipsum\n",
                array("lineLength"=>30, "cutOnlyAtSpace"=>false)
            ),
            array(
                array(
                    "long.description" => "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    "exact.max.length" => "Lorem ips",
                    "big.first.word" => "Loremipsumdolor sit amet, consectetur dipiscing elit.",
                    "with.long.space" => "Praesent a quam a mauris   tempus   placerat. Sed bibendum     mattis erat, quis consequat ",
                ),
                "long.description = Lorem\\\n".
                "ipsum dolor sit amet,\\\n".
                "consectetur adipiscing elit.\n".
                "exact.max.length = Lorem ips\n".
                "big.first.word = Loremipsumdolor\\\n".
                "sit amet, consectetur\\\n".
                "dipiscing elit.\n".
                "with.long.space = Praesent a\\\n".
                "quam a mauris   tempus\\\n".
                "\\s placerat. Sed bibendum   \\s\\\n".
                "mattis erat, quis consequat\\s\n"
                ,
                array("lineLength"=>28, "cutOnlyAtSpace"=>true)
            ),
            array(
                array(
                    "long.description" => "Lorem ipsu#m dolor \nsit amet, consectetur adipiscing elit.",
                    "exact.max.length" => "Lorem i#psum"
                ),
                "# This is \n".
                "# a header comment\n".
                "long.description = Lorem ipsu\\#\\\n".
                "m dolor \\nsit amet, consectetu\\\n".
                "r adipiscing elit.\n".
                "exact.max.length = Lorem i\\#ps\\\n".
                "um\n",
                array("lineLength"=>30, "cutOnlyAtSpace"=>false, "headerComment"=> "This is \na header comment")
            ),
            array(
                array(
                    "super.long" => "<p>Un email vous a été envoyé à l'adresse indiquée. Si vous ne le voyez pas ".
                        "dans votre boite au lettre d'ici quelques minutes, vérifiez votre dossier de".
                        " spam. Si vous n'avez pas reçu le mail, peut-être que le nom d'utilisateur ".
                        "et l'adresse que vous avez donné ne sont pas valides. Contactez ".
                        "l'administrateur du site web.</p> <p>Cet email contient un lien vers une ".
                        "page qui vous permettra de changer de mot de passe. Ce lien est valable 48 ".
                        "heures.</p> <p>Tant que vous n'avez pas changé le mot de passe, l'ancien ".
                        "reste valable.</p>",
                ),
                "super.long = <p>Un email vous a été envoyé à l'adresse indiquée. Si vous ne le voyez pas dans votre boite au lettre d'ici quelques minutes, vérifiez votre dossier de spam. Si vous n'avez pas reçu le\\\n".
                "mail, peut-être que le nom d'utilisateur et l'adresse que vous avez donné ne sont pas valides. Contactez l'administrateur du site web.</p> <p>Cet email contient un lien vers une page qui vous\\\n".
                "permettra de changer de mot de passe. Ce lien est valable 48 heures.</p> <p>Tant que vous n'avez pas changé le mot de passe, l'ancien reste valable.</p>\n",
                array("lineLength"=>200, "cutOnlyAtSpace"=>true)
            ),



        );
    }

    /**
     * @dataProvider getPropertiesContent
     */
    public function testWriterString($properties, $expected, $options){
        $props = new Properties($properties);
        $writer = new Writer ();
        $result = $writer->writeToString($props, $options);

        $this->assertEquals($expected, $result);
    }
}

src/Parser.php
<?php
/**
 * @author     Laurent Jouanneau
 * @contributor Gerald Croes, Julien Issler, Yannick Le Guédart, Dominique Papin
 * @copyright 2005-2018 Laurent Jouanneau, 2001-2005 CopixTeam, 2008 Julien Issler, 2008 Yannick Le Guédart, 2008 Dominique Papin
 *
 * @link       https://jelix.org
 * @licence    MIT
 */
namespace Jelix\PropertiesFile;

/**
 * reads a properties file
 */
class Parser
{
    /**
     */
    public function __construct()
    {
    }


    /**
     * @param string $fileName  the file to parse
     * @param Properties $properties the container when readed properties will be stored
     * @param string $charset  the charset used into the file
     * @throws \Exception
     * @throws SyntaxException
     */
    public function parseFromFile($fileName, Properties $properties, $charset = 'UTF-8')
    {
        $f = @fopen($fileName, 'r');

        if ($f === false) {
            throw new \Exception('Cannot load the properties file ' . $fileName);
        }
        $nexLine = function () use ($f) {
            if (feof($f)) {
                fclose($f);
                return false;
            }
            return fgets($f);
        };

        try {
            $this->readContent($properties, $nexLine, $charset);
        } catch (SyntaxException $e) {
            fclose($f);
            $msg = sprintf($e->getMessage(), 'file '.$fileName);
            throw new SyntaxException($msg, $e->getCode(), $e);
        }
    }

    /**
     * @param Properties $properties
     * @param callable $nextLine
     * @param string $charset
     */
    protected function readContent(Properties $properties, callable $nextLine, $charset)
    {
        $utf8Mod = ($charset == 'UTF-8') ? 'u' : '';
        $unbreakablespace = ($charset == 'UTF-8') ? utf8_encode(chr(160)) : chr(160);
        $escapedChars = array('\#', '\n', '\w', '\S', '\s');
        $unescape = array('#', "\n", ' ', $unbreakablespace, ' ');
        $multiline = false;
        $lineNumber = 0;
        $key = '';

        while (($line = $nextLine()) !== false) {
            ++$lineNumber;
            $line = rtrim($line);

            if ($multiline) {
                // the current line is the part of the value of the previous readed property
                if (preg_match("/^\s*(.*)\s*(\\\\?)$/U".$utf8Mod, $line, $match)) {
                    $multiline = ($match[2] == '\\');
                    if (strlen($match[1])) {
                        $sp = preg_split('/(?<!\\\\)\#/', $match[1], -1, PREG_SPLIT_NO_EMPTY);
                        $properties[$key] .= ' '.str_replace($escapedChars, $unescape, trim($sp[0]));
                    } else {
                        $properties[$key] .= ' ';
                    }
                } else {
                    throw new SyntaxException('Syntax error in properties %s, line '.$lineNumber);
                }
            } elseif (preg_match("/^\s*(.+)\s*=\s*(.*)\s*(\\\\?)$/U".$utf8Mod, $line, $match)) {
                // we got a key=value
                $key = $match[1];
                $multiline = ($match[3] == '\\');
                $sp = preg_split('/(?<!\\\\)\#/', $match[2], -1, PREG_SPLIT_NO_EMPTY);
                if (count($sp)) {
                    $value = trim($sp[0]);
                } else {
                    $value = '';
                }
                $properties[$key] = str_replace($escapedChars, $unescape, $value);
            } elseif (preg_match("/^\s*(\#.*)?$/", $line, $match)) {
                // ok, just a comment, let's ignore it
            } else {
                throw new SyntaxException('Syntax error in properties content line '.$lineNumber);
            }
        }
    }
}


src/Properties.php
<?php
/**
 * @author      Laurent Jouanneau
 * @copyright   2018 Laurent Jouanneau
 *
 * @link        https://jelix.org
 * @licence     MIT
 */
namespace Jelix\PropertiesFile;

/**
 * Container for properties readed from a properties file
 * @package Jelix\PropertiesFile
 */
class Properties implements \ArrayAccess
{

    /**
     * @var string[] list of key/values
     */
    protected $properties = array();

    /**
     * Properties constructor.
     * @param string[] $properties initial properties
     */
    public function __construct($properties = array())
    {
        $this->properties = $properties;
    }

    /**
     * Get a property
     *
     * @param string $key
     * @return null|string
     */
    public function get($key)
    {
        if (isset($this->properties[$key])) {
            return $this->properties[$key];
        }
        return null;
    }

    /**
     * Return all properties
     *
     * @return string[]
     */
    public function getAllProperties()
    {
        return $this->properties;
    }

    /**
     * Gets an iterator on all properties
     *
     * @return \ArrayIterator the iterator
     */
    public function getIterator()
    {
        return new \ArrayIterator($this->properties);
    }

    /**
     * Set a property
     *
     * @param string $key
     * @param string $value
     */
    public function set($key, $value)
    {
        $this->properties[$key] = $value;
    }

    // --------------------------- ArrayAccess interface

    public function offsetSet($key, $value)
    {
        if (is_null($key)) {
            $this->properties[] = $value;
        } else {
            $this->properties[$key] = $value;
        }
    }

    public function offsetExists($key)
    {
        return isset($this->properties[$key]);
    }

    public function offsetUnset($key)
    {
        unset($this->properties[$key]);
    }

    public function offsetGet($key)
    {
        return isset($this->properties[$key]) ? $this->properties[$key] : null;
    }
}


src/Writer.php
<?php
/**
 * @author      Laurent Jouanneau
 * @copyright   2018 Laurent Jouanneau
 *
 * @link        https://jelix.org
 * @licence     MIT
 */
namespace Jelix\PropertiesFile;

class Writer {

    const DEFAULT_OPTIONS = array(
        "lineLength" => 120,
        "spaceAroundEqual" => true,
        "headerComment" => "",
        "removeTrailingSpace" => false,
        "cutOnlyAtSpace" => true,
        "encoding" => "UTF-8"
    );

    function writeToFile(Properties $properties, $fileName, $options = array()) {

        $options = array_merge(self::DEFAULT_OPTIONS, $options);

        $f = @fopen($fileName, 'w');

        if ($f === false) {
            throw new \Exception('Cannot open the properties file ' . $fileName);
        }

        $lineWriter = function($line) use ($f) {
            fwrite($f, $line);
        };

        $this->writeContent($lineWriter, $properties, $options);
        fclose($f);
    }

    function writeToString(Properties $properties, $options = array()) {

        $options = array_merge(self::DEFAULT_OPTIONS, $options);

        $o = new \StdClass;
        $o->content = '';

        $lineWriter = function($line) use ($o) {
            $o->content .= $line;
        };

        $this->writeContent($lineWriter, $properties, $options);
        return $o->content;
    }

    protected function writeContent(callable $writer, Properties $properties, $options) {

        if ($options["headerComment"]) {
            $writer('# '.str_replace("\n", "\n# ",$options["headerComment"])."\n");
        }

        $equal = ($options["spaceAroundEqual"]? ' = ': '=');

        foreach($properties->getIterator() as $key => $value) {
            $line = $key . $equal;


            $value = mb_ereg_replace("#", "\\#", $value);
            $value = mb_ereg_replace(utf8_encode(chr(160)), "\\S", $value);
            $value = mb_ereg_replace("\n", "\\n", $value);
            if ($options["removeTrailingSpace"]) {
                $value = rtrim($value, " ");
            }

            $startLen = mb_strlen($line, $options['encoding']);
            $valueLen = mb_strlen($value, $options['encoding']);
            if ($valueLen + $startLen > $options['lineLength']) {
                $value = $this->chunkSplit($value,
                    $options['lineLength'] - $startLen,
                    $options['lineLength'],
                    $options['cutOnlyAtSpace'],
                    $options['encoding']);
            }
            $value = mb_ereg_replace(" $", "\\s", $value);
            $value = mb_ereg_replace("^ ", "\\s", $value);
            $line .= $value;

            $writer($line."\n");
        }
    }


    protected function chunkSplit($string, $firstLineLength, $lineLength, $cutOnlyAtSpace, $encoding) {
        $valueLen = mb_strlen($string, $encoding);
        $start = 0;
        $line = "";
        if ($valueLen > $firstLineLength) {
            list ($cut, $cutLength) = $this->cutString($string, $start, $firstLineLength, $cutOnlyAtSpace, $encoding);
            $line .= $cut;
            $valueLen -= $cutLength;
            $start += $cutLength;
        }

        while ($valueLen > $lineLength) {
            list ($cut, $cutLength) = $this->cutString($string, $start, $lineLength, $cutOnlyAtSpace, $encoding);
            $line .= $cut;
            $valueLen -= $cutLength;
            $start += $cutLength;
        }
        if ($valueLen > 0) {
            $line .= mb_substr($string, $start, $valueLen, $encoding);
        }
        $line = preg_replace('/(\\n )/mu', "\n\\s", $line);
        return preg_replace('/( \\\\\\n)/mu', "\\s\\\n", $line);
    }

    protected function cutString($string, $start, $lineLength, $cutOnlyAtSpace, $encoding) {

        $cut = mb_substr($string, $start, $lineLength, $encoding);
        $length = mb_strlen($cut, $encoding);

        if ($cutOnlyAtSpace) {
            $cut2 = preg_replace('/(\s+?\S+)?$/u', '', mb_substr($string, $start, $lineLength+1, $encoding));
            $length2 = mb_strlen($cut2, $encoding);
            if ($length < $length2) {
                $lastchar = mb_substr($string, $start+$lineLength, 1, $encoding);
                if ($lastchar != ' ') {
                    $cut = $cut2.preg_replace('/^(\S+)(.*)/u', '${1}', mb_substr($string, $start+$lineLength+1, null, $encoding));
                    $length = mb_strlen($cut, $encoding);
                }
            }
            else if ($length > $length2) {
                $cut = $cut2;
                $length = $length2;
            }
        }
        else {
            if (substr($cut, -1) == '\\') { //escape character
                $cut = mb_substr($string, $start, $lineLength+1, $encoding);
                $length = $lineLength + 1;
            }
        }

        if (mb_substr($string, $start+$length, 1, $encoding) == ' ') {
            $length++;
        }
        return array($cut."\\\n", $length);
    }

}